xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,305 @@
+
+func (gsb *gracefulSwitchBalancer) resolveNow(bal balancer.Balancer, opts resolver.ResolveNowOptions) {
+	// If the resolver specifies an update with a whole new type of policy,
+	// there is no reason to forward a ResolveNow call from the balancer being
+	// switched from, as the most recent update from the Resolver does not
+	// concern the balancer being switched from.
+	gsb.mu.Lock()
+	if bal == gsb.balancerCurrent && gsb.balancerPending != nil {
+		gsb.mu.Unlock()
+		return
+	}
+	gsb.mu.Unlock()
+	gsb.cc.ResolveNow(opts)
+}
+
+func (gsb *gracefulSwitchBalancer) SwitchTo(childType string) error {
+	if gsb.closed.HasFired() {
+		// logger?
+		return errBalancerClosed
+	}
+
+	// Now, the question is whether we compare to previous type here or always
+	// switch to no matter what? If so you don't even need to persist
+	// recentChildType Used to be nested in the if buildPolicy block of
+	// UpdateClientConnState(). If you want to compare, have this be in that if
+	// conditional (compared to a persisted recentChildType), if not, that if
+	// conditional is determined by the caller.
+
+	builder := balancer.Get(childType)
+	if builder == nil {
+		return balancer.ErrBadResolverState
+		// return fmt.Errorf("balancer of type %v not supported", lbCfg.ChildBalancerType) // Maybe make this ErrBadResolverState as well?
+	}
+	gsb.mu.Lock()
+	ccw := &clientConnWrapper{
+		ClientConn: gsb.cc,
+		gsb:        gsb,
+	}
+	newBalancer := builder.Build(ccw, gsb.bOpts)
+	ccw.bal = newBalancer
+	if gsb.balancerCurrent == nil {
+		gsb.balancerCurrent = newBalancer
+	} else {
+		gsb.pendingState = balancer.State{}
+		// Clean up resources here that are from a previous pending lb.
+		for sc, sb := range gsb.scToSubBalancer {
+			if sb == gsb.balancerPending {
+				delete(gsb.scToSubBalancer, sc)
+			}
+		}
+		gsb.balancerPending = newBalancer
+	}
+	gsb.mu.Unlock()
+	return nil
+}
+
+// "Simply forwarding the update to one of the children"
+func (gsb *gracefulSwitchBalancer) UpdateClientConnState(state balancer.ClientConnState) error {
+	if gsb.closed.HasFired() {
+		// logger?
+		return errBalancerClosed
+	}
+	gsb.mu.Lock()
+	var balToUpdate balancer.Balancer
+	if gsb.balancerPending != nil {
+		balToUpdate = gsb.balancerPending
+	} else {
+		balToUpdate = gsb.balancerCurrent
+	}
+	gsb.mu.Unlock()
+	balToUpdate.UpdateClientConnState(state)
+}
+
+func (gsb *gracefulSwitchBalancer) ResolverError(err error) {
+	if gsb.closed.HasFired() {
+		// Logger?
+		return
+	}
+	gsb.mu.Lock()
+	currentBalancerToUpdate := gsb.balancerCurrent
+	pendingBalancerToUpdate := gsb.balancerPending
+	gsb.mu.Unlock()
+	if currentBalancerToUpdate != nil {
+		currentBalancerToUpdate.ResolverError(err)
+	}
+	if pendingBalancerToUpdate != nil {
+		pendingBalancerToUpdate.ResolverError(err)
+	}
+}
+
+func (gsb *gracefulSwitchBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	if gsb.closed.HasFired() {
+		// Logger
+		return
+	}
+	gsb.mu.Lock()
```

Comment:
here and in many other places: is the decision to not defer the `Unlock` intentional? And if so, why?